File: src/tp_src/learning_gem5/part2/TPHelloObject.py
from m5.params import *
from m5.SimObject import SimObject

class TPHelloObject(SimObject):
	type = 'TPHelloObject'
	cxx_header = "tp_src/learning_gem5/part2/tp_hello_object.hh"
	cxx_class = "gem5::TPHelloObject"

	time_to_wait = Param.Latency("Time before firing event")
	number_of_fires = Param.Int(1, "Number of times to fire the event before goodbye")
	goodbye_object = Param.TPGoodbyeObject("A tp_goodbye object")

class TPGoodbyeObject(SimObject):
	type = 'TPGoodbyeObject'
	cxx_header = "tp_src/learning_gem5/part2/tp_goodbye_object.hh"
	cxx_class = "gem5::TPGoodbyeObject"

	buffer_size = Param.MemorySize(
		'1kB',
		"Size of buffer to fill with goodbye")
	#tp_write_bandwith = Param.MemoryBandwith(
	#	'10MB/s',
	#	"Bandwith to fill the buffer")

	write_bandwidth = Param.MemoryBandwidth('100MB/s', "Bandwidth to fill "
                                            "the buffer")

File: src/tp_src/learning_gem5/part2/tp_hello_object.cc
#include "tp_src/learning_gem5/part2/tp_hello_object.hh"
#include "debug/TPHello.hh"

#include <iostream>

namespace gem5 {

TPHelloObject::TPHelloObject(const TPHelloObjectParams &params) :
    SimObject(params), 
	event([this]{processEvent();}, name()),
	goodbye(params.goodbye_object),
	myName(params.name),
	latency(params.time_to_wait),
	timesLeft(params.number_of_fires) {
    DPRINTF(TPHello, "Created the hello object with the name %s\n", myName);
	panic_if(!goodbye, "TPHelloObject must have a non-null TPGoodbyeObject");
}

void TPHelloObject::startup() {
	schedule(event, latency);
}

void TPHelloObject::processEvent() {
	timesLeft--;
	DPRINTF(TPHello, "Hello world! Processing the event! %d left\n", timesLeft);
	
	if (timesLeft <= 0) {
		DPRINTF(TPHello, "Done firing!\n");
		goodbye->sayGoodbye(myName);
	} else {
		schedule(event, curTick() + latency);
	}
}

} // namespace gem5
